keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,282 @@
+
+Since we do this step on the client-side we have access to the full functionality of the `install` command including installing operators from the file system. This will come very handy during the development and debugging which arguably becomes more complex with dependencies.
+
+##### Server-Side
+
+Upon receiving a new operator Instance with dependencies KUDO mangers workflow engine will:
+
+1. Build a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph) by transitively expanding top-level `deploy` plan using operator-tasks as vertices, and their execution order (`a` needs `b` to be installed first) as edges
+2. Perform cycle detection and fail if circular dependencies found. We could additionally run this check on the client-side as part of the `kudo package verify` command to improve the UX
+3. If we haven't found any cycles, start executing the top-level `deploy` plan. When encountering an operator-task, apply the corresponding `Instance` resource. Here it is the same as for any other resource that we create: we check if it is healthy and if not, end current plan execution and "wait" for it to become healthy. KUDO manager already has a [health check](https://github.com/kudobuilder/kudo/blob/master/pkg/engine/health/health.go#L78) for `Instance` resources implemented.
+
+Let's take a look at an example. Here is a simplified operator `AA` with a few dependencies:
+
+```text
+AA
+├── BB
+│   ├── EE
+│   │   ├── H
+│   │   └── I
+│   ├── F
+│   └── GG
+│       ├── J
+│       └── K
+├── CC
+│   ├── L
+│   └── M
+└── D
+
+Legend:
+- Operators and operator-tasks are marked with double letters e.g. 'AA' or `BB`
+- other tasks are marked with single letters e.g. 'D'
+- direct children of an operator are the 'deploy' plan steps e.g. for 'AA' deploy steps are 'BB', 'CC' and 'D'
+```
+
+In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.
+
+The instance controller (IC) then starts with the execution of the top-level `deploy` plan of the operator `AA`. The first task is the `BB` operator-task. When executing it, IC creates the `Instance-BB` resource and ends current reconciliation. Next, IC notices new `Instance-BB` resource, starts new reconciliation, and executes the  `deploy` plan of the operator `BB` which then creates `Instance-EE` resource. This way we are basically performing the depth-first search for the dependency graph, executing each vertex in the right order e.g. `EE` has to be healthy before `BB` deploy plan can continue with the next step `F`.
```

Comment:
One thing that Terraform does that I really like is add a root node to that dependency graph as the "permanent" start point. This is handy so that you can have multiple "roots" (still non-cyclical!) that you can execute in parallel. Do we want to include that concept here?

Reply:
In the model above, the one root is the `AA` `deploy` plan. However, it becomes interesting if it has multiple _parallel_ phases, each then having an operator-task. In this case, we might actually build _multiple_ graphs and execute them in parallel. I'll keep that in the back of my head and revisit once we get to the implementation. Good point 👍